NDTensors/src/lib/GradedAxes/src/unitrangedual.jl

```diff
@@ -92,6 +105,14 @@ BlockArrays.blockaxes(a::UnitRangeDual) = blockaxes(nondual(a))
 BlockArrays.blockfirsts(a::UnitRangeDual) = label_dual.(blockfirsts(nondual(a)))
 BlockArrays.blocklasts(a::UnitRangeDual) = label_dual.(blocklasts(nondual(a)))
 BlockArrays.findblock(a::UnitRangeDual, index::Integer) = findblock(nondual(a), index)
+
+blocklabels(a::UnitRangeDual) = dual.(blocklabels(nondual(a)))
+
+gradedisequal(::UnitRangeDual, ::AbstractGradedUnitRange) = false
+gradedisequal(::AbstractGradedUnitRange, ::UnitRangeDual) = false
```

Comment:
I guess I could see the argument that one might define this such that `gradedisequal(gradedrange([U1(-1) => 2]), dual(gradedrange([U1(1) => 2])))` returns `true`, is there a reason to choose the definition one way or another?

Additionally, I would hope that these definitions wouldn't be needed for dual ranges, and instead we could make the generic definition "just work". For example, the generic definition is:
```julia
  blockisequal(a1, a2) && (blocklabels(a1) == blocklabels(a2))
```
so as long as `blockisequal` and `blocklabels` are defined for dual ranges, that definition would work and we wouldn't need specialized definition for dual unit ranges.

Reply:
There is a good reason indeed. For abelian groups groups it does not matter, however a spin 1/2 or its self conjugate dual are not the same (same space, different basis). They must not be compared as equal, `dual` matters so it is not enough to just check `blockisequal` and `blocklabels`.

Reply:
Hmm, then I am not following the definition you wrote above, since for abelian groups shouldn't it potentially output `true`, while the definition above always outputs `false`?

And wouldn't a generic definition:
```julia
  blockisequal(a1, a2) && (blocklabels(a1) == blocklabels(a2))
```
catch the non-abelian case, since in that case the `blocklabels` wouldn't be equal? Maybe I'm not understanding something.

@emstoudenmire any opinions on this?

Reply:
If you only work with abelian groups, you may decide not to explicitly implement the distinction between dual and non-dual spaces, and to take dual label wise (similar to `flip`) whenever a dual is required. This is not possible for SU(2), where an explicit distinction between dual and non-dual is required. Note that just checking `blocklabels` is not enough as any SU(2) irrep is self-dual. Consequently we have to explicitly implement dual and check for it in equality.

It is _possible_ to have different code paths for abelian and non-abelian groups, but I think it would lead to more complex code and subtle debugging issues for a final user.

Indeed the use case I have in mind for `gradedisequal` is to check whether two axes can be contracted. This is only mathematically well defined if one axis is the dual of the other.  Imagine you write a tensor network contraction code where you try to contract two non dual axes. You may first test and debug it for `gradedrange([U1(0)=>2])`: with your suggestion, there would no error and no way to spot the issue. Then you may try the very same contraction with `gradedrange([SU2(1/2)=>1])` as input: there you should get an error (and if not the results will be wrong). I think it would result in easier debugging and better user experience not to treat abelian groups differently and to impose the same constraints in every cases.

Reply:
I see, that helps to explain the definitions you are making. I don't think we should use `gradedisequal` for that purpose, maybe we should have a `can_contract` function, which maybe uses `gradedisequal` internally but also checks that you are contracting a dual with a non-dual space.

I was picturing that `gradedisequal` is by definition:
```julia
  blockisequal(a1, a2) && (blocklabels(a1) == blocklabels(a2))
```
independent of whether or not the spaces are dual, if they are abelian or non-abelian, etc.

Reply:
That sounds fine, I think we're both in agreement about having different kinds of equality and we're just debating names and code design at this point.

My suggestion would be something like:
```julia
function labelled_isequal(a1::AbstractBlockedUnitRange, a2::AbstractBlockedUnitRange)
  return blockisequal(a1, a2) && (blocklabels(a1) == blocklabels(a2))
end

function space_isequal(a1::AbstractUnitRange, a2::AbstractUnitRange)
  return (isdual(a1) && isdual(a2)) && labelled_isequal(a1, a2)
end
```
though I'm open to suggestions about the names.

Part of my point is also about the code style, specifically I would advocate here and elsewhere for using more generic function signatures with generic implementations that are built out of other generic functions, instead of "overtyping" the method signatures. I'm not sure if you use that style because you think it has better performance since it pushes more things to compile time, but I don't think that is a concern for small functions like this and Julia's compiler is pretty good at hoisting things to compile time using techniques like union splitting and constant propagation.

Reply:
I split `gradedisequal` into `labelled_isequal` and `space_isequal`. I also define `blocklabels(::AbstractUnitRange) = nothing` as a default to be able to use them in generic code.

Reply:
Thanks, I'll resolve this thread, though note that I had some suggestions for how to define `blocklabels(::AbstractUnitRange)` in https://github.com/ITensor/ITensors.jl/pull/1363#discussion_r1779659318.